First portion of the draft for AI Local Inference. (#284)
jvanderaa/jvanderaa.github.io: content/posts/2026/ai_local_inference_memory/index.md

---
date: 2026-03-28
slug: local-inference-memory
categories:
- automation
- ai
title: "AI: Local Models Memory"
summary: >
  TODO: Update
toc: true
tags:
  - ai
  - local_ai
  - openclaw
author: jvanderaa
coverAlt: AI Memory Bursting
coverCaption:
params:
  showComments: true
---

> For a long while I was getting it all wrong with my set up of AI local inference. I thought that for each sub-agent that I would spin up on my OpenClaw that the system would use that amount of VRAM. I've been running [Qwen3.5-122B-A10B](https://huggingface.co/unsloth/Qwen3.5-122B-A10B-GGUF) at 6bit. This model uses 101GB in the loading of the model. So I thought if I had two sub-agents using this model that I would need 202GB of RAM. **I was wrong**.

Reviewer: **Grammar/clarity fixes:**

```suggestion
For a long while I was getting it all wrong with my setup of AI local inference. I thought that for each sub-agent that I would spin up on my OpenClaw, the system would use that amount of VRAM. I've been running [Qwen3.5-122B-A10B](https://huggingface.co/unsloth/Qwen3.5-122B-A10B-GGUF) at 6-bit quantization. This model uses 101GB when loading the weights. So I thought if I had two sub-agents using this model, I would need 202GB of RAM. **I was wrong**.
```

Changes: "set up" → "setup" (noun form), added comma after "OpenClaw", "6bit" → "6-bit quantization" (clearer for readers), "in the loading of the model" → "when loading the weights" (tighter).

> So what I was doing was loading multiple models into my system and pointing different sub-agents at different models. Where my heavy coding agent would use a much lighter than the model that I would be using for reasoning and research.

Reviewer: **Incomplete sentence - missing object:**

```suggestion
So what I was doing was loading multiple models into my system and pointing different sub-agents at different models. My heavy coding agent would use a much lighter model than the one I was using for reasoning and research.
```

The original is missing its object: "would use a much lighter **[what?]** than the model" needs "model" added. Also "Where my heavy" is a sentence fragment.

![Wrong memory footprint](image.png)

## Introduction to Weights

> I finally went ahead and asked an AI system (I forget if I talked to Harry or Gemini or ChatGPT or Claude about this). I asked the simple question, am I doing this right in that I need to have multiple models running for each sub-agent to be able to have each of the sub-agents working simultaneously, and the quick answer was that I was wrong. I was then learned that the bulk of the memory numbers that are quote are [AI Model Weights](https://www.ntia.gov/programs-and-initiatives/artificial-intelligence/open-model-weights-report/background). This is the information that is loaded and used to respond to prompts that are provided.

Reviewer: **Grammar fixes (multiple issues):**

```suggestion
I finally went ahead and asked an AI system (I forget which one - Harry, Gemini, ChatGPT, or Claude). I asked the simple question: do I need to have multiple models running for each sub-agent to work simultaneously? The quick answer was no. I then learned that the bulk of the memory numbers you see quoted are for [AI Model Weights](https://www.ntia.gov/programs-and-initiatives/artificial-intelligence/open-model-weights-report/background) - the trained parameters that get loaded into memory and used to respond to prompts.
```

Fixes:
- "I was then learned" → "I then learned"
- "numbers that are quote are" → "numbers you see quoted are for"
- Tightened the run-on sentence into shorter, clearer statements

> That being, that the weights are loaded in, and then there is a KV-Cache, which is the calculation of the context that is processed. So the system will use this KV-Cache as part of the processing. And this is dynamic in nature that grows with each prompt that is processed.

Reviewer: **Clarity - add a concrete example to land the aha moment:**

```suggestion
The weights are loaded once into memory. On top of that, each active request gets a KV-cache - a working memory that stores the computed attention keys and values for the current context. The KV-cache is dynamic: it grows with each token processed and is allocated per-slot, not per-model. So running 4 agents against one model means one copy of weights + 4 small KV-caches, not 4 copies of the model.
```

The original is technically correct but vague. Adding the concrete example makes the point immediately clear.

Here is what is actually happening on the system:

![AI Memory Allocations](image-1.png)

> ## Summary

Reviewer: **Tighten the summary - currently a bit repetitive:**

```suggestion
## Summary

The technology road is quite the journey. This might be a "yeah, you didn't know that?" moment for some, but I didn't, and I've learned. Since consolidating all my sub-agents onto the same local model, the performance across the board has improved significantly - even for the agents that were previously on smaller models. The system is more stable, and I'm getting better results.

If I had this misunderstanding, there's a non-zero chance someone else does too. Hope this helps.

-Josh
```

Cuts repetition ("I have learned" appears twice, "the best part is" doesn't add). Keeps Josh's voice but punchier.
